app/utils.py

```diff
@@ -154,29 +162,141 @@ def setup_logger(name, level=logging.INFO):
-        if column.nullable is False and column.name != 'id':
-            # strip _id from category_id
-            name = column.name.replace('_id', '')
-            required.append(name)
+
+        # There are only required parameters for POSTing new resources.
+        if request.method == 'POST':
+            if column.nullable is False and column.name != 'id':
+                # strip _id from category_id
+                name = column.name.replace('_id', '')
+                required.append(name)
+
+        # Type Validation.
+        col_type = column.type.python_type
+        field = json.get(column.name.replace('_id', ''))
+
+        if field is not None:
+
+            # Category is a foreign key (int) relation but we only accept string input.
+            # The provided category will get mapped to this resource later on.
+            if column.name == 'category_id':
+                col_type = str
+
+            # Allow String columns to accept integers.
+            if col_type is str:
+                if type(field) in [int, float]:
+
+                    # Do not allow integers as urls
+                    if column.name.replace('_id', '') != 'url':
+                        continue
+
+            # Allow Bool columns to accept strings that are variations "true" or "false"
+            if col_type is bool:
+                if type(field) is str and field.lower() in ["false", "true"]:
+                    continue
+
+            # Validate that fields are of the correct type.
+            if col_type is not type(field):
+                validation_errors['errors'].append(
+                    generate_error(ValidationError.TypeError,
+                                   column.name.replace('_id', ''),
+                                   received=type(field).__name__,
+                                   expected=col_type.__name__)
+                )
+
+    # Special case for Languages - Must be list and all elements must be str.
+    langs = json.get('languages')
+    if langs:
+        if type(langs) is not list or not all(map(lambda x: type(x) is str, langs)):
+            validation_errors['errors'].append(
+                generate_error(ValidationError.TypeError,
+                               "languages",
+                               received=type(langs).__name__,
+                               expected=list.__name__)
+            )
 
     for prop in required:
         if json.get(prop) is None:
-            validation_errors['errors']['missing_params'].append(prop)
-
-    if validation_errors['errors']['missing_params']:
+            validation_errors['errors'].append(
+                generate_error(ValidationError.MissingParamError, prop))
+
+    url = json.get("url")
+    if url is not None:
+
+        # If there is an existing resource with the requested url
+        # and it isn't the Resource we are trying to update, return an error.
+        resource = Resource.query.filter_by(url=url).first()
```

Comment:
Isn't this unnecessary now that we're catching the `IntegrityError` when we try to actually create/update the record in the database? It would be better to just have the one query to insert it and return the `ValidationError.IntegrityError` from the catch block in the routes file, right?

Reply:
Good catch.

Reply:
Upon further discussion in slack, @Digbigpig and I decided that getting rid of it is premature optimization. We should still catch the `IntegrityError` and return a 422 in that case, but we should stop it before it gets that far. If later we decide that we want to implement caching to prevent making 2 db calls in subsequent requests, we'll consider that when we find that it's necessary. "Premature optimization is the root of all evil"